### PR TITLE
Added benchmarking infrastructure

### PR DESCRIPTION
**Pull request by theoden8**
I've added some benchmarks to measure, compare and keep track of the performance.

### CODE REVIEW

File: bench/benchmark.hpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+#include <iostream>
+#include <string>
+#include <unistd.h>
+#include <utility>
+
+template <typename ScalarT>
+ScalarT *new_data(size_t size, bool initialized = true) {
+  ScalarT *v = new ScalarT[size];
+  if (initialized) {
+#pragma omp parallel for
```
Comment (Ruyk) [inline bench/benchmark.hpp:15]: Please remove the OpenMP

File: bench/clblas_benchmark.cpp [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+
+  BENCHMARK_FUNCTION(asum_bench) {
+    UNPACK_PARAM;
+    double flops;
+    {
+      ScalarT vr;
+      MemBuffer<ScalarT, CL_MEM_WRITE_ONLY> buf1(context, size);
+      MemBuffer<ScalarT, CL_MEM_READ_ONLY> bufr(context, &vr, 1);
+      cl_mem scratch = clCreateBuffer(context, CL_MEM_READ_WRITE,
+                                      size * sizeof(ScalarT), NULL, NULL);
```
Comment (Ruyk) [inline bench/clblas_benchmark.cpp:104]: Please be aware of the different flags for creating buffers: https://www.khronos.org/registry/OpenCL/sdk/1.2/docs/man/xhtml/clCreateBuffer.html

File: bench/clblast_benchmark.cpp [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+      });
+    }
+    return flops;
+  }
+
+  BENCHMARK_FUNCTION(axpy_bench) {
+    UNPACK_PARAM;
+    double flops;
+    {
+      ScalarT alpha(2.4367453465);
```
Comment (Ruyk) [inline bench/clblast_benchmark.cpp:40]: create a global constant for your random number

File: bench/clblast_benchmark.cpp [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+                               context._queue(), &events[0]);
+        clblast::Asum<ScalarT>(size, bufr.dev(), 0, buf2.dev(), 0, 1,
+                               context._queue(), &events[1]);
+        clblast::Dot<ScalarT>(size, bufr.dev(), 1, buf1.dev(), 0, 1, buf2.dev(),
+                              0, 1, context._queue(), &events[2]);
+        clblast::Nrm2<ScalarT>(size, bufr.dev(), 2, buf1.dev(), 0, 1,
+                               context._queue(), &events[3]);
+        clblast::Amax<ScalarT>(size, buf_i.dev(), 0, buf1.dev(), 0, 1,
+                               context._queue(), &events[4]);
+        clWaitForEvents(5, events);
```
Comment (Ruyk) [inline bench/clblast_benchmark.cpp:245]: sizeof(events)

File: bench/clwrap.hpp [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+  void release() {
+    clReleaseCommandQueue(command_queue);
+    clReleaseContext(context);
+    is_active = false;
+  }
+
+  operator cl_context() const { return context; }
+
+  cl_command_queue *_queue() { return &command_queue; }
```
Comment (Ruyk) [inline bench/clwrap.hpp:64]: why do you need this one?

File: bench/clwrap.hpp [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+template <typename ScalarT, int Options = CL_MEM_READ_WRITE>
+class MemBuffer {
+ public:
+  static constexpr const bool to_write =
+      (Options == CL_MEM_READ_WRITE || Options == CL_MEM_WRITE_ONLY);
+  static constexpr const bool to_read =
+      (Options == CL_MEM_READ_WRITE || Options == CL_MEM_READ_ONLY);
+
+ private:
+  Context &context;
```
Comment (Ruyk) [inline bench/clwrap.hpp:82]: why a reference?

File: bench/syclblas_benchmark.cpp [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+        q.wait_and_throw();
+      });
+    }
+    release_data(v1);
+    release_data(v2);
+    return flops;
+  }
+
+  BENCHMARK_FUNCTION(iamax_bench) {
+    UNPACK_PARAM;
```
Comment (Ruyk) [inline bench/syclblas_benchmark.cpp:138]: Can this be a function instead of a macro?

**Review by Ruyk — CHANGES_REQUESTED** [review]
Some comments, seems you have done a lot of work, well done!

File: bench/clwrap.hpp (hunk 5 above)
Comment (theoden8) [reply bench/clwrap.hpp:64]: Clblast wants cl_command_queue * as an argument

File: bench/clblast_benchmark.cpp (hunk 4 above)
Comment (theoden8) [reply bench/clblast_benchmark.cpp:245]: I am not sure it is a part of the standard. There is an example of how to use sizeof(block), but I have not found an explicit statement.

File: bench/clblast_benchmark.cpp (hunk 3 above)
Comment (theoden8) [reply bench/clblast_benchmark.cpp:40]: It is type-dependent, only can make a macro out of it.

File: bench/clblas_benchmark.cpp (hunk 2 above)
Comment (theoden8) [inline bench/clblas_benchmark.cpp]: I would ask for help with understanding which of them are important for the benchmarks.

File: bench/benchmark.hpp (hunk 1 above)
Comment (theoden8) [reply bench/benchmark.hpp:15]: Done

File: bench/syclblas_benchmark.cpp (hunk 7 above)
Comment (theoden8) [reply bench/syclblas_benchmark.cpp:138]: Yes, but I wanted to make the same template arguments, return value and parameters which would be defined and used through benchmarks.hpp

**Pull request closed without merging**